service/history/chasm_engine.go

```diff
@@ -241,14 +256,125 @@ func (e *ChasmEngine) ReadComponent(
 	return readFn(chasmContext, component)
 }
 
+// PollComponent waits until the supplied predicate is satisfied when evaluated against the
+// component identified by the supplied component reference. If there is no error, it returns (ref,
+// nil) where ref is a component reference identifying the state at which the predicate was
+// satisfied. It's possible that multiple state transitions (multiple notifications) occur between
+// predicate checks, therefore the predicate must be monotonic: if it returns true at execution
+// state transition s it must return true at all transitions t > s. It is an error if execution
+// transition history is (after reloading from persistence) behind the requested ref, or if the ref
+// is inconsistent with execution transition history. Thus when the predicate function is evaluated,
+// it is guaranteed that the execution VT >= requestRef VT. opts are currently ignored.
+// PollComponent subscribes to execution-level notifications. Suppose that an execution consists of
+// one component A, and A has subcomponent B. Subscribers interested only in component B may be
+// woken up unnecessarily (and thus evaluate the predicate unnecessarily) due to changes in parts of
+// A that do not also belong to B.
 func (e *ChasmEngine) PollComponent(
 	ctx context.Context,
-	entityRef chasm.ComponentRef,
-	predicateFn func(chasm.Context, chasm.Component) (any, bool, error),
-	operationFn func(chasm.MutableContext, chasm.Component, any) error,
+	requestRef chasm.ComponentRef,
+	monotonicPredicate func(chasm.Context, chasm.Component) (bool, error),
 	opts ...chasm.TransitionOption,
-) (newEntityRef []byte, retError error) {
-	return nil, serviceerror.NewUnimplemented("PollComponent is not yet supported")
+) (retRef []byte, retError error) {
+
+	var ch <-chan struct{}
+	var unsubscribe func()
+	defer func() {
+		if unsubscribe != nil {
+			unsubscribe()
+		}
+	}()
+
+	checkPredicateOrSubscribe := func() ([]byte, error) {
+		_, executionLease, err := e.getExecutionLease(ctx, requestRef)
+		if err != nil {
+			return nil, err
+		}
+		defer executionLease.GetReleaseFn()(nil)
+
+		ref, err := e.predicateSatisfied(ctx, monotonicPredicate, requestRef, executionLease)
+		if err != nil {
+			if errors.Is(err, consts.ErrStaleState) {
+				err = serviceerror.NewUnavailable("please retry")
```

Comment:
I need a better error message here, but I haven't decided what.
> "request data ahead of persisted data"

?

Also, should it be `PollActivityExecution` that converts this to `serviceerror`?